basic_pitch/data/tf_example_deserialization.py

```diff
@@ -0,0 +1,697 @@
+) -> Tuple[tf.Tensor, Dict[str, tf.Tensor], Dict[str, tf.Tensor]]:
+    return (
+        # mix the audio snippets
+        tf.math.reduce_mean(a, axis=0),
+        # annotations are the max per bin - active frames stay active
+        {
+            "onset": tf.math.reduce_max(target["onset"], axis=0),
+            "contour": tf.math.reduce_max(target["contour"], axis=0),
+            "note": tf.math.reduce_max(target["note"], axis=0),
+        },
+        # weights are the minimum - if an annotation is missing in one, we should set the weights to zero
+        {
+            "onset": tf.math.reduce_min(w["onset"], axis=0),
+            "contour": tf.math.reduce_min(w["contour"], axis=0),
+            "note": tf.math.reduce_min(w["note"], axis=0),
+        },
+    )
+
+
+def transcription_dataset(
+    file_generator: Callable[[], Iterator[str]], n_samples_per_track: int, random_seed: bool
+) -> tf.data.Dataset:
+    """
+    `fpaths_in` is a list of .tfrecords files
+    return a tf.Dataset with the following fields (as tuple):
+        - audio (shape AUDIO_N_SAMPLES, 1)
+        - {'contours': contours, 'notes': notes, 'onsets': onsets}
+    contours has shape (ANNOT_N_FRAMES, N_FREQ_BINS_CONTOURS)
+    notes and onsets have shape: (ANNOT_N_FRAMES, N_FREQ_BINS_NOTES)
+    """
+    ds = tf.data.Dataset.from_generator(file_generator, output_types=tf.string, output_shapes=())
+    ds = tf.data.TFRecordDataset(ds)
+    ds = ds.map(parse_transcription_tfexample, num_parallel_calls=tf.data.AUTOTUNE)
+    ds = ds.filter(is_not_bad_shape)
+    ds = ds.map(
+        lambda file_id, source, audio_wav, notes_indices, notes_values, onsets_indices, onsets_values, contours_indices, contours_values, notes_onsets_shape, contours_shape: (  # noqa: E501
+            file_id,
+            source,
+            tf.audio.decode_wav(
+                audio_wav,
+                desired_channels=AUDIO_N_CHANNELS,
+                desired_samples=-1,
+                name=None,
+            ),
+            sparse2dense(notes_values, notes_indices, notes_onsets_shape),
+            sparse2dense(onsets_values, onsets_indices, notes_onsets_shape),
+            sparse2dense(contours_values, contours_indices, contours_shape),
+        ),
+        num_parallel_calls=tf.data.AUTOTUNE,
+    )
+    ds = ds.map(reduce_transcription_inputs)
+    ds = ds.map(get_sample_weights, num_parallel_calls=tf.data.AUTOTUNE)
+    ds = ds.flat_map(
+        lambda a, o, c, n, ow, cw, nw, m: get_transcription_chunks(
+            a, o, c, n, ow, cw, nw, n_samples_per_track, random_seed
+        )
+    )
+    ds = ds.filter(is_not_all_silent_annotations)  # remove examples where all annotations are zero
+    ds = ds.map(to_transcription_training_input)
+    ds = ds.apply(tf.data.experimental.ignore_errors(log_warning=True))  # failsafe so training doesn't stop
+    return ds
+
+
+def parse_transcription_tfexample(
+    serialized_example: tf.train.Example,
+) -> Tuple[
```

Comment:
This is uh...kinda ugly. Was this from the original code?

Reply:
yes